### PR TITLE
Fix resolving links in sharded directories on gateway

### PR DESCRIPTION
**Pull request by magik6k**
Fixes https://github.com/ipfs/go-ipfs/issues/5270

### CODE REVIEW

File: plugin/loader/load_linux.go [hunk 1]
```diff
@@ -1,3 +1,5 @@
+// +build !noplugin
```
Comment (Stebalien) [inline plugin/loader/load_linux.go]: Looks like a `git add .`.

**Comment by magik6k** [comment]
should be ready

File: path/resolver/resolver.go [hunk 2]
```diff
@@ -69,19 +69,28 @@ func (r *Resolver) ResolveToLastNode(ctx context.Context, fpath path.Path) (ipld
...
+			next, err := lnk.GetNode(ctx, r.DAG)
 			if err != nil {
 				return nil, nil, err
 			}
 			nd = next
 			p = rest
+			continue
+		}
+
+		val, rest, err := nd.Resolve(p)
```
Comment (whyrusleeping) [inline path/resolver/resolver.go]: is 'rest' not used?

File: path/resolver/resolver.go [hunk 3]
```diff
@@ -69,19 +69,28 @@ func (r *Resolver) ResolveToLastNode(ctx context.Context, fpath path.Path) (ipld
 	}
 
 	for len(p) > 0 {
-		val, rest, err := nd.Resolve(p)
-		if err != nil {
-			return nil, nil, err
-		}
+		lnk, rest, err := r.ResolveOnce(ctx, r.DAG, nd, p)
+		if lnk != nil {
```
Comment (whyrusleeping) [inline path/resolver/resolver.go]: maybe we could clean this loop up a little bit by pulling everything below this if statement out of the for loop, then reversing the logic to 'break' if `lnk == nil`

File: path/resolver/resolver.go [hunk 4]
```diff
@@ -69,25 +69,42 @@ func (r *Resolver) ResolveToLastNode(ctx context.Context, fpath path.Path) (ipld
 	}
 
 	for len(p) > 0 {
-		val, rest, err := nd.Resolve(p)
+		lnk, rest, err := r.ResolveOnce(ctx, r.DAG, nd, p)
+		if lnk == nil {
```
Comment (Stebalien) [inline path/resolver/resolver.go]: This'll drop the error, as far as I can tell.

Comment (whyrusleeping) [reply path/resolver/resolver.go]: Ah, interesting. Can we make sure to document that explicitly? At a glance, it looks wrong, so let's make sure nobody comes along and 'fixes' it

Comment (Stebalien) [reply path/resolver/resolver.go]: We should probably remove the error check below in that case.

Comment (magik6k) [reply path/resolver/resolver.go]: It can technically still be set at that point, but that would be a rather weird edge case

**Pull request merged**